File: backend/app/routers/admin_analytics.py
"""Admin analytics API endpoints using PostgreSQL - no authentication, security through obscurity."""

from typing import List, Dict, Any
from fastapi import APIRouter, HTTPException

from app.db.postgres import get_postgres_pool
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/admin-analytics", tags=["admin-analytics"])


@router.get("/overview", response_model=Dict[str, Any])
async def get_analytics_overview():
    """Get high-level analytics overview."""
    try:
        pool = await get_postgres_pool()
        
        async with pool.acquire() as conn:
            # Count totals from PostgreSQL tables
            total_users = await conn.fetchval("SELECT COUNT(*) FROM users")
            total_quiz_attempts = await conn.fetchval("SELECT COUNT(*) FROM quiz_results")
            total_feedback = await conn.fetchval("SELECT COUNT(*) FROM flashcard_feedback")
            total_bookmarks = await conn.fetchval("SELECT COUNT(*) FROM bookmarks")
            
            # Get feedback breakdown
            likes = await conn.fetchval("SELECT COUNT(*) FROM flashcard_feedback WHERE rating = 1")
            dislikes = await conn.fetchval("SELECT COUNT(*) FROM flashcard_feedback WHERE rating = -1")
        
        overview = {
            "total_users": total_users or 0,
            "total_sessions": 0,  # Sessions moved to quiz_sessions
            "completed_sessions": 0,
            "total_quiz_attempts": total_quiz_attempts or 0,
            "total_feedback": total_feedback or 0,
            "feedback_likes": likes or 0,
            "feedback_dislikes": dislikes or 0,
            "total_bookmarks": total_bookmarks or 0
        }
        
        logger.info("Retrieved admin analytics overview")
        return overview
        
    except Exception as e:
        logger.error(f"Error getting analytics overview: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get overview: {str(e)}")


@router.get("/quiz-results", response_model=List[Dict[str, Any]])
async def get_quiz_performance_summary():
    """Get aggregated quiz performance by deck."""
    try:
        pool = await get_postgres_pool()
        
        query = """
            SELECT 
                course_id,
                deck_id,
                COUNT(*) as total_attempts,
                AVG(percentage) as average_score,
                MAX(percentage) as highest_score,
                MIN(percentage) as lowest_score,
                AVG(time_taken) as average_time,
                AVG(total_questions) as total_questions_avg
            FROM quiz_results
            GROUP BY course_id, deck_id
            ORDER BY total_attempts DESC
        """
        
        async with pool.acquire() as conn:
            rows = await conn.fetch(query)
        
        formatted_stats = []
        for row in rows:
            formatted_stats.append({
                "course_id": row["course_id"],
                "deck_id": row["deck_id"],
                "total_attempts": row["total_attempts"],
                "average_score_percentage": round(row["average_score"] or 0, 2),
                "highest_score_percentage": round(row["highest_score"] or 0, 2),
                "lowest_score_percentage": round(row["lowest_score"] or 0, 2),
                "average_time_seconds": round(row["average_time"] or 0, 2),
                "average_total_questions": round(row["total_questions_avg"] or 0, 1)
            })
        
        logger.info(f"Retrieved quiz performance summary for {len(formatted_stats)} decks")
        return formatted_stats
        
    except Exception as e:
        logger.error(f"Error getting quiz performance summary: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get quiz performance: {str(e)}")


@router.get("/flashcard-feedback/summary", response_model=List[Dict[str, Any]])
async def get_flashcard_feedback_summary():
    """Get like/dislike counts per flashcard."""
    try:
        pool = await get_postgres_pool()
        
        query = """
            SELECT 
                course_id,
                deck_id,
                flashcard_index,
                COUNT(*) as total_feedback,
                SUM(CASE WHEN rating = 1 THEN 1 ELSE 0 END) as likes,
                SUM(CASE WHEN rating = -1 THEN 1 ELSE 0 END) as dislikes
            FROM flashcard_feedback
            GROUP BY course_id, deck_id, flashcard_index
            ORDER BY total_feedback DESC
        """
        
        async with pool.acquire() as conn:
            rows = await conn.fetch(query)
        
        formatted_stats = []
        for row in rows:
            net_score = (row["likes"] or 0) - (row["dislikes"] or 0)
            
            formatted_stats.append({
                "course_id": row["course_id"],
                "deck_id": row["deck_id"],
                "flashcard_index": row["flashcard_index"],
                "flashcard_identifier": f"{row['course_id']}:{row['deck_id']}:{row['flashcard_index']}",
                "total_feedback": row["total_feedback"],
                "likes": row["likes"] or 0,
                "dislikes": row["dislikes"] or 0,
                "net_score": net_score
            })
        
        logger.info(f"Retrieved flashcard feedback summary for {len(formatted_stats)} flashcards")
        return formatted_stats
        
    except Exception as e:
        logger.error(f"Error getting flashcard feedback summary: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get feedback summary: {str(e)}")


@router.get("/flashcard-feedback/details", response_model=List[Dict[str, Any]])
async def get_flashcard_feedback_details():
    """Get raw feedback log with user/session info."""
    try:
        pool = await get_postgres_pool()
        
        query = """
            SELECT 
                id, user_id, session_id, course_id, deck_id,
                flashcard_index, rating, created_at
            FROM flashcard_feedback
            ORDER BY created_at DESC
        """
        
        async with pool.acquire() as conn:
            rows = await conn.fetch(query)
        
        formatted_feedback = []
        for row in rows:
            formatted_feedback.append({
                "feedback_id": str(row["id"]),
                "user_id": row["user_id"],
                "session_id": row["session_id"],
                "course_id": row["course_id"],
                "deck_id": row["deck_id"],
                "flashcard_index": row["flashcard_index"],
                "flashcard_identifier": f"{row['course_id']}:{row['deck_id']}:{row['flashcard_index']}",
                "rating": row["rating"],
                "rating_text": "like" if row["rating"] == 1 else "dislike",
                "created_at": row["created_at"]
            })
        
        logger.info(f"Retrieved {len(formatted_feedback)} feedback details")
        return formatted_feedback
        
    except Exception as e:
        logger.error(f"Error getting flashcard feedback details: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get feedback details: {str(e)}")


@router.get("/users", response_model=List[Dict[str, Any]])
async def get_all_users_summary():
    """Get all user summaries with activity stats."""
    try:
        pool = await get_postgres_pool()
        
        query = """
            SELECT 
                firebase_uid, created_at, last_active,
                total_decks_studied, total_quiz_attempts
            FROM users
            ORDER BY last_active DESC
        """
        
        async with pool.acquire() as conn:
            rows = await conn.fetch(query)
        
        formatted_users = []
        for row in rows:
            formatted_users.append({
                "user_id": row["firebase_uid"],
                "created_at": row["created_at"],
                "last_active": row["last_active"],
                "total_decks_studied": row["total_decks_studied"] or 0,
                "total_quiz_attempts": row["total_quiz_attempts"] or 0
            })
        
        logger.info(f"Retrieved {len(formatted_users)} users summary")
        return formatted_users
        
    except Exception as e:
        logger.error(f"Error getting users summary: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get users summary: {str(e)}")

File: frontend/src/api/adminAnalytics.js
/**
 * API functions for admin analytics (no authentication required)
 */

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';
const API_ENDPOINT = `${API_BASE_URL}/api/v1/admin-analytics`;

/**
 * Get analytics overview with high-level statistics
 * @returns {Promise<Object>} Overview statistics
 */
export async function getAnalyticsOverview() {
  try {
    const response = await fetch(`${API_ENDPOINT}/overview`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get analytics overview: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting analytics overview:', error);
    throw error;
  }
}

/**
 * Get all study sessions with details
 * @returns {Promise<Array>} Array of all study sessions
 */
export async function getAllSessions() {
  try {
    const response = await fetch(`${API_ENDPOINT}/sessions`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get all sessions: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting all sessions:', error);
    throw error;
  }
}

/**
 * Get quiz performance summary by deck
 * @returns {Promise<Array>} Array of quiz performance statistics by deck
 */
export async function getQuizPerformanceSummary() {
  try {
    const response = await fetch(`${API_ENDPOINT}/quiz-results`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get quiz performance summary: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting quiz performance summary:', error);
    throw error;
  }
}

/**
 * Get flashcard feedback summary with like/dislike counts
 * @returns {Promise<Array>} Array of flashcard feedback statistics
 */
export async function getFlashcardFeedbackSummary() {
  try {
    const response = await fetch(`${API_ENDPOINT}/flashcard-feedback/summary`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get flashcard feedback summary: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting flashcard feedback summary:', error);
    throw error;
  }
}

/**
 * Get detailed flashcard feedback log
 * @returns {Promise<Array>} Array of individual feedback entries
 */
export async function getFlashcardFeedbackDetails() {
  try {
    const response = await fetch(`${API_ENDPOINT}/flashcard-feedback/details`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get flashcard feedback details: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting flashcard feedback details:', error);
    throw error;
  }
}

/**
 * Get all users summary with activity stats
 * @returns {Promise<Array>} Array of all user summaries
 */
export async function getAllUsersSummary() {
  try {
    const response = await fetch(`${API_ENDPOINT}/users`, {
      method: 'GET',
    });

    if (!response.ok) {
      throw new Error(`Failed to get users summary: ${response.status}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Error getting users summary:', error);
    throw error;
  }
}

/**
 * Get filtered sessions by date range
 * @param {Date} startDate - Start date for filtering
 * @param {Date} endDate - End date for filtering
 * @returns {Promise<Array>} Filtered array of sessions
 */
export async function getSessionsByDateRange(startDate, endDate) {
  try {
    const allSessions = await getAllSessions();
    
    return allSessions.filter(session => {
      const sessionDate = new Date(session.session_start_time);
      return sessionDate >= startDate && sessionDate <= endDate;
    });
  } catch (error) {
    console.error('Error getting sessions by date range:', error);
    throw error;
  }
}

/**
 * Get top performing flashcards by net feedback score
 * @param {number} limit - Number of top flashcards to return
 * @returns {Promise<Array>} Array of top performing flashcards
 */
export async function getTopPerformingFlashcards(limit = 10) {
  try {
    const feedbackSummary = await getFlashcardFeedbackSummary();
    
    // Sort by net score (likes - dislikes) in descending order
    const topFlashcards = feedbackSummary
      .sort((a, b) => b.net_score - a.net_score)
      .slice(0, limit);
    
    return topFlashcards;
  } catch (error) {
    console.error('Error getting top performing flashcards:', error);
    throw error;
  }
}

/**
 * Get worst performing flashcards by net feedback score
 * @param {number} limit - Number of worst flashcards to return
 * @returns {Promise<Array>} Array of worst performing flashcards
 */
export async function getWorstPerformingFlashcards(limit = 10) {
  try {
    const feedbackSummary = await getFlashcardFeedbackSummary();
    
    // Sort by net score (likes - dislikes) in ascending order
    const worstFlashcards = feedbackSummary
      .sort((a, b) => a.net_score - b.net_score)
      .slice(0, limit);
    
    return worstFlashcards;
  } catch (error) {
    console.error('Error getting worst performing flashcards:', error);
    throw error;
  }
}

/**
 * Get user activity statistics
 * @returns {Promise<Object>} User activity statistics
 */
export async function getUserActivityStats() {
  try {
    const [overview, sessions] = await Promise.all([
      getAnalyticsOverview(),
      getAllSessions()
    ]);
    
    // Calculate additional stats
    const completedSessions = sessions.filter(session => session.is_completed).length;
    const avgSessionDuration = sessions
      .filter(session => session.study_duration_seconds)
      .reduce((sum, session) => sum + session.study_duration_seconds, 0) / sessions.length || 0;
    
    return {
      ...overview,
      completion_rate: overview.total_sessions > 0 ? (completedSessions / overview.total_sessions) * 100 : 0,
      average_session_duration_minutes: Math.round(avgSessionDuration / 60)
    };
  } catch (error) {
    console.error('Error getting user activity stats:', error);
    throw error;
  }
}
